pkg/ipmi/ipmi.go

```diff
@@ -205,47 +204,102 @@ func fdSet(fd uintptr, p *syscall.FdSet) {
 	p.Bits[fd/64] |= 1 << (uint(fd) % 64)
 }
 
-func (i *IPMI) sendrecv(req *req) ([]byte, error) {
-	addr := systemInterfaceAddr{
+func ioctlSetReq(fd, name uintptr, req *req) error {
+	_, _, err := unix.Syscall(unix.SYS_IOCTL, fd, name, uintptr(unsafe.Pointer(req)))
+	runtime.KeepAlive(req)
+	if err != 0 {
+		return err
+	}
+	return nil
+}
+
+func ioctlGetRecv(fd, name uintptr, recv *recv) error {
+	_, _, err := unix.Syscall(unix.SYS_IOCTL, fd, name, uintptr(unsafe.Pointer(recv)))
+	runtime.KeepAlive(recv)
+	if err != 0 {
+		return err
+	}
+	return nil
+}
+
+// SendRecvBasic sends the IPMI message, receives the response, and returns the
+// response data. This is recommended for use unless the user must be able to
+// specify the data pointer and length on their own.
+func (i *IPMI) SendRecvBasic(netfn NetFn, cmd Command, data []byte) ([]byte, error) {
+	msg := Msg{
+		Netfn:   netfn,
+		Cmd:     cmd,
+		Data:    unsafe.Pointer(&data[0]),
+		DataLen: uint16(len(data)),
+	}
+	return i.SendRecv(msg)
+}
+
+// SendRecv sends the IPMI message, receives the response, and returns the
+// response data.
+func (i *IPMI) SendRecv(msg Msg) ([]byte, error) {
+	addr := &systemInterfaceAddr{
 		addrType: _IPMI_SYSTEM_INTERFACE_ADDR_TYPE,
 		channel:  _IPMI_BMC_CHANNEL,
 	}
+	req := &req{
+		addr:    addr,
+		addrLen: uint32(unsafe.Sizeof(addr)),
+		msgid:   rand.Int63(),
+		msg:     msg,
+	}
 
-	req.addr = &addr
-	req.addrLen = uint32(unsafe.Sizeof(addr))
 	if err := ioctlSetReq(i.Fd(), _IPMICTL_SEND_COMMAND, req); err != nil {
 		return nil, err
 	}
 
 	set := &syscall.FdSet{}
 	fdSet(i.Fd(), set)
-	time := &syscall.Timeval{
+	timeval := &syscall.Timeval{
 		Sec:  _IPMI_OPENIPMI_READ_TIMEOUT,
 		Usec: 0,
 	}
-	if _, err := syscall.Select(int(i.Fd()+1), set, nil, nil, time); err != nil {
+	if _, err := syscall.Select(int(i.Fd()+1), set, nil, nil, timeval); err != nil {
 		return nil, err
 	}
 
-	recv := &recv{}
-	recv.addr = &systemInterfaceAddr{}
-	recv.addrLen = uint32(unsafe.Sizeof(addr))
 	buf := make([]byte, _IPMI_BUF_SIZE)
-	recv.msg.data = unsafe.Pointer(&buf[0])
-	recv.msg.dataLen = _IPMI_BUF_SIZE
-	if err := ioctlGetRecv(i.Fd(), _IPMICTL_RECEIVE_MSG, recv); err != nil {
-		return nil, err
+	recvMsg := Msg{
+		Data:    unsafe.Pointer(&buf[0]),
+		DataLen: _IPMI_BUF_SIZE,
+	}
+	recv := &recv{
+		addr:    req.addr,
+		addrLen: req.addrLen,
+		msg:     recvMsg,
+	}
+
+	t := time.After(timeout)
+	for {
+		if err := ioctlGetRecv(i.Fd(), _IPMICTL_RECEIVE_MSG, recv); err != nil {
```

Comment:
Why is this being repeated now? Shouldn't the select above mean that there is data here?

Reply:
I thought if the msg id doesn't match we'd need to repeat it - that may be wrong, should I remove it?

Reply:
is there some C reference code I could read, or a spec? i'm not sure.

Reply:
I was looking at http://openipmi.sourceforge.net/IPMI.pdf pages 29-31 - it talks about using the msg to correlate cmds and responses, but doesn't necessarily mean that we need to repeat the ioctl, so I'm not sure